util/gate/gate.go

```diff
@@ -13,23 +13,36 @@
 
 package gate
 
-import "context"
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
 
 // A Gate controls the maximum number of concurrently running and waiting queries.
 type Gate struct {
-	ch chan struct{}
+	ch           chan struct{}
+	waitDuration prometheus.Counter
 }
 
 // New returns a query gate that limits the number of queries
 // being concurrently executed.
-func New(length int) *Gate {
+func New(length int, waitDuration prometheus.Counter) *Gate {
```

Comment:
I think the metric parameter here should be a pointer :thinking:

Reply:
I tried to rethink the suggestion once again, and it seems to have worked out. You can review it in this commit: https://github.com/prometheus/prometheus/pull/13565/commits/f223bfbaa411bd6a3c428e2e64e800593061cda2.

If the quantity and nature of changes are not entirely acceptable, everting the commit would be convenient.

PS Sorry if my previous comment is too silly :)

Reply:
Sorry, I made this comment just based on how we do similar "passing of common metrics", for example [here](https://github.com/prometheus/prometheus/blob/main/storage/remote/queue_manager.go#L443) where we pass a pointer to a struct of metrics. I still think the pointer is the right way to go :thinking:

Reply:
Understood, anyway, it's implemented with a pointer now, as you suggested, and it looks not bad IMO. That was an interesting indication.
Then I'll be waiting for further review and approvals.
Thanks.